### PR TITLE
Compute speaker embeddings in batch for the LSTM  Speaker Encoder and Compute embeddings/ finding chars using config file.

### PR DESCRIPTION
**Pull request by Edresson**

### CODE REVIEW

File: TTS/tts/datasets/preprocess.py [hunk 1]
```diff
@@ -270,16 +272,20 @@ def libri_tts(root_path, meta_files=None):
...
         with open(meta_file, "r") as ttf:
             for line in ttf:
                 cols = line.split("\t")
-                wav_file = os.path.join(_root_path, cols[0] + ".wav")
-                text = cols[1]
+                file_name = cols[0]
+                speaker_name, chapter_id, *_ = cols[0].split("_")
+                _root_path = os.path.join(root_path, f"{speaker_name}/{chapter_id}")
+                wav_file = os.path.join(_root_path, file_name + ".wav")
+                text = cols[2]
```
Comment (Edresson) [inline TTS/tts/datasets/preprocess.py:288]: There was a bug in the libriTTS data loader. Perhaps, for this reason, the training results of the TTS models were often not very good in this dataset. Column [1] of the libriTTS contains the non-normalized transcript. For example, century is not spelled out as "XXVI". On the other hand, column [2] is normalized with the transcript spelled out "twenty six."

File: TTS/bin/compute_embeddings.py [hunk 2]
```diff
@@ -1,73 +1,45 @@
...
+    "config_dataset_path",
     type=str,
-    default="",
-    help="Target dataset to pick a processor from TTS.tts.dataset.preprocess. Necessary to create a speakers.json file.",
+    help="Path to dataset config file.",
 )
+parser.add_argument("output_path", type=str, help="path for output speakers.json and/or speakers.npy.")
 parser.add_argument("--use_cuda", type=bool, help="flag to set cuda.", default=True)
-parser.add_argument("--separator", type=str, help="Separator used in file if CSV is passed for data_path", default="|")
+parser.add_argument("--save_npy", type=bool, help="flag to set cuda.", default=False)
```
Comment (erogol) [inline TTS/bin/compute_embeddings.py:31]: the help msg here is wrong

File: TTS/speaker_encoder/models/lstm.py [hunk 3]
```diff
@@ -70,24 +71,32 @@ def inference(self, x):
             d = torch.nn.functional.normalize(d, p=2, dim=1)
         return d
 
-    def compute_embedding(self, x, num_frames=160, overlap=0.5):
```
Comment (erogol) [inline TTS/speaker_encoder/models/lstm.py:73]: why you removed `overlap` and add `num_eval`? Do they do the same thing?

File: TTS/bin/compute_embeddings.py [hunk 4]
```diff
@@ -1,73 +1,45 @@
 import argparse
```
Comment (erogol) [inline TTS/bin/compute_embeddings.py:1]: In general, it is better to use the speaker_manager in this script to reduce the maintenance cost. 

Your changes here demand changes in the speaker_manager any ways.

File: TTS/bin/compute_embeddings.py [hunk 5]
```diff
@@ -100,11 +72,20 @@
 
 if speaker_mapping:
     # save speaker_mapping if target dataset is defined
-    if ".json" not in args.output_path:
+    if '.json' not in args.output_path and '.npy' not in args.output_path:
```
Comment (erogol) [inline TTS/bin/compute_embeddings.py:75]: Do `.json` and `.npy` files have the same content?

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: If you use the speaker_manager it already provides a function to compute embeddings from multiple instances.

File: TTS/tts/datasets/preprocess.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ def split_dataset(items):
     return items[:eval_split_size], items[eval_split_size:]
 
 
-def load_meta_data(datasets, eval_split=True):
+def load_meta_data(datasets, eval_split=True, ignore_generated_eval=False):
```
Comment (erogol) [inline TTS/tts/datasets/preprocess.py:40]: `eval_split` should already let people choose to generate a eval split or not.  Why someone needs to ignore the eval split.

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/bin/compute_embeddings.py (hunk 5 above)
Comment (Edresson) [reply TTS/bin/compute_embeddings.py:75]: Yes, but the numpy takes up much less space. When working with large files this may be a better option than .json.

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: Can you explain what `num_eval` does here?  Does it set the overlap based on the expected number of evaluations?

File: TTS/bin/compute_embeddings.py (hunk 4 above)
Comment (erogol) [reply TTS/bin/compute_embeddings.py:1]: update it :)

File: TTS/bin/compute_embeddings.py (hunk 5 above)
Comment (erogol) [reply TTS/bin/compute_embeddings.py]: Yeah makes sense.  But then you need to update the speaker manager to handle npy too to keep the coder running after your PR

**Comment by erogol** [comment]
I'll review once again tomorrow sorry for the delay.

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (Edresson) [reply TTS/speaker_encoder/models/lstm.py:73]: That's right, basically, if we want 10 samples regardless of the number of frames we will have 10 samples. That way we guarantee an x usage of memory all the time. It is more controllable than overlapping. And the state-of-the-art papers are using so I implemented it that way.

File: TTS/tts/datasets/preprocess.py (hunk 6 above)
Comment (Edresson) [reply TTS/tts/datasets/preprocess.py:40]: I think as the speaker embeddings will be used for model training, during training you will always need the eval. So I don't see a case where the user doesn't need to extract the embeddings from the eval. But if you still want I can add it as in argparser.

File: TTS/bin/compute_embeddings.py (hunk 4 above)
Comment (Edresson) [reply TTS/bin/compute_embeddings.py:1]: I updated :)

File: TTS/bin/compute_embeddings.py (hunk 5 above)
Comment (Edresson) [reply TTS/bin/compute_embeddings.py]: I removed the .npy save at this point, as in most cases space is not an issue.

File: TTS/tts/datasets/preprocess.py (hunk 6 above)
Comment (erogol) [reply TTS/tts/datasets/preprocess.py:40]: I think it is better to add it to argparse and remove this from the function.

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: But then we use the same number of frames regardless of the length of the audio right?

Comment (Edresson) [reply TTS/speaker_encoder/models/lstm.py:73]: yeah, exactly. It's better for memory consumption control.

File: TTS/tts/datasets/preprocess.py (hunk 6 above)
Comment (Edresson) [reply TTS/tts/datasets/preprocess.py:40]: Ignoring the generated eval is necessary as in some cases some samples are lost. If we don't ignore it in some cases it won't calculate all embeddings causing errors during training. Therefore, if eval_split is not passed in config.json, compute from all and avoid splitting. I cannot remove this until I have a better solution. The current split_dataset loses some samples during the split.

Comment (Edresson) [reply TTS/tts/datasets/preprocess.py]: I kept the parameter: ignore_generated_eval and added as argparser the eval_split. Please review again.

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: but it also prevents us to compute better features for longer audio clips. Am I wrong ?

Comment (Edresson) [reply TTS/speaker_encoder/models/lstm.py:73]: not exactly because we randomly choose windows on the audio, so it can get all the parts of the audio.

File: TTS/tts/datasets/__init__.py [hunk 7]
```diff
@@ -30,7 +30,7 @@ def split_dataset(items):
     return items[:eval_split_size], items[eval_split_size:]
 
 
-def load_meta_data(datasets, eval_split=True):
+def load_meta_data(datasets, eval_split=True, ignore_generated_eval=False):
```
Comment (erogol) [inline TTS/tts/datasets/__init__.py:33]: Honestly, I still don't understand why do you need ìgnore_generated_eval`


If you set eval_split=False you do the same thing as setting `ignore_generated_eval=true`

File: TTS/speaker_encoder/models/lstm.py (hunk 3 above)
Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: I think it is better to keep both options. Maybe we can separate these two into two different functions. One with overlap and one with a number of frames. I think computing with overlap is more useful especially at the inference time.

Comment (Edresson) [reply TTS/speaker_encoder/models/lstm.py:73]: I'm not so sure about that. During some tests in the past, I compared the EER of the old implementation (overlapping) with this new one, both with the default parameters and the EER with the new implementation was better (ie embedding was better). This can be explained because taking the window randomly overlaps in different parts of the audio. If you still want I can add the two new functions, but I prefer to use this new one in extract_embeddings.

Comment (erogol) [reply TTS/speaker_encoder/models/lstm.py:73]: ok makes sense. Let's keep it as it is now.

File: TTS/tts/datasets/__init__.py (hunk 7 above)
Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: yes but we need to compute the eval_split=True otherwise the eval specified in config.json will be ignored and we need these embeddings during training. Here we need to prevent generating an eval (ignore_generated_eval=True) and at the same time compute the eval  specified in config.json

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: But they don't run at the same time. It either uses the specified eval split or computes an eval split. But they all run if `eval_split`is set `True`. 

So your flag does not change anything.

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: I think you still don't understand the purpose of the flag. The flag works for its purpose which is to prevent the dataset from being split into training and validation (call of the split_dataset function) during the extraction of spectrograms and speaker embeddings.

if eval_split=True and ignore_generated_eval=True and meta_file_val=None (ie a validation csv was not specified in config) this will ignore the split_dataset function call (thus it will not split the dataset into training and validation).

To extract spectrograms and speaker embeddings, it is necessary to ignore the split_dataset function call. This function can remove samples during dataset split between training and validation (in the case of multispeaker training). If samples are removed when extracting embeddings the training will break (since the embedding of that sample will be missing).

We cannot simply pass eval_split=False because for the speaker embeddings extraction the validation set is necessary (since validation will be used in training, if we do not extract the speaker embeddings from these samples the training will break !). 

So if we extract the speaker embeddings with eval_split=False and the user specifies validation in the configuration file ( meta_file_val != None). When the user starts training with these speaker embeddings, the training will break during the validation, because the speakers.json does not have the speaker embeddings from the validation set.

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: if you set `eval_split=false` there is no eval split and you don't need to prevent it since it does not exist. 

So eval_split=True and ignore_generated_eval=True and meta_file_val=None is equal to eval_split=False

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: If you set eval_split=False, there is no eval split and all the dataset is regarded as training data and all the d-vectors will be extracted anyways.

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: Why don't you set meta_file_val to None in the compute_speaker_embedding rather than adding an argument to save the situation in somewhere else ?

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: Indeed. However we have no control (in the extraction script) over whether meta_file_val is specified (meta_file_val!=None) or not (meta_file_val=None) . So we need a flag, in case the user doesn't specify an eval ( meta_file_val==None) we ignore the problem of possible loss of samples caused by splitting the dataset (split_dataset function call) and this is where the flag was made !

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: > Why don't you set meta_file_val to None in the compute_speaker_embedding rather than adding an argument to save the situation in somewhere else ?

It has datasets that have validation and training in separate CSVs. This way we would force the user to change config.json just to extract the speaker embeddings. Forcing him to put the validation file as training in config.json. And this can cause confusion! The purpose of the compute_embeddings script is for the user to use the config.json used for training and just run the script getting the result and a speakers.json ready for training. The flag is the easiest way to enable this without extra work on the user's part!

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: So this does not solve the problem in `compute_embeddings.py`?
 
```python
train_items, eval_items = load_meta_data(dataset_config, eval_split=False)
data_items = train_items + eval_items
```

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: In my branch it's ok: https://github.com/Edresson/Coqui-TTS/blob/dev/TTS/bin/compute_embeddings.py#L35

make sure you are not looking at the wrong commit.

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: I am confused :) ?
What do you mean it's ok? 
Does the change I suggested above solve the issue or not?

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: oh ok sorry my bad.

 No it's not solve. As I explained above we need to ignore the generated validation. Because if we use the split_dataset function to split the training and validation this can cause loss of samples when the dataset is multispeaker.

Comment (erogol) [reply TTS/tts/datasets/__init__.py:33]: you don't lose any samples above. We merge eval and train samples.

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: you still don't understand the problem! It is necessary to skip the eval split generation when the user does not provide a CSV for validation !. The split_dataset function that splits the dataset in training and validation can in a multispeaker scenario remove samples (please see here: https://github.com/Edresson/Coqui-TTS/blob/d906fea08cf96c95d78fa01cdc935fc6a87c6685/TTS/tts/datasets___init .py#L28). So we need to ignore the split_dataset function call and not generate an eval_split if the user doesn't specify it. I've already added eval+training, but that doesn't solve it because the split_dataset function can remove samples in multispeaker scenarios.

Comment (Edresson) [reply TTS/tts/datasets/__init__.py:33]: ok, I found the problem. Apparently it was some old bug in the function now everything works fine (after merges). If this happens again I can see how to solve it and send another PR. So at the moment I tried to reproduce the problem after the merge and it doesn't happen anymore.,

File: TTS/bin/compute_embeddings.py [hunk 8]
```diff
@@ -1,80 +1,41 @@
...
-import torch
 from tqdm import tqdm
 
-from TTS.config import BaseDatasetConfig, load_config
-from TTS.speaker_encoder.utils.generic_utils import setup_model
+from TTS.config import load_config
 from TTS.tts.datasets import load_meta_data
 from TTS.tts.utils.speakers import SpeakerManager
-from TTS.utils.audio import AudioProcessor
+
```
Comment (erogol) [inline TTS/bin/compute_embeddings.py:10]: Can you add an example call as docstring here so that people know how to call it?

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/bin/compute_embeddings.py (hunk 8 above)
Comment (Edresson) [reply TTS/bin/compute_embeddings.py:11]: I did this, but the test failed on python3.6.  Any idea how to solve it?

Comment (erogol) [reply TTS/bin/compute_embeddings.py:11]: let me run the test again
